icu4c/source/tools/genrb/reslist.cpp

```diff
@@ -1390,20 +1388,24 @@ SRBRoot::compactKeys(UErrorCode &errorCode) {
         ++keys;  /* skip the NUL */
     }
     keys = fKeys + fKeysBottom;
-    for (; i < keysCount; ++i) {
+    while (i < keysCount) {
         int32_t keyOffset = static_cast<int32_t>(keys - fKeys);
-        while (!fIsPoolBundle && keysInUse.count(keyOffset) == 0) {
+        if (!fIsPoolBundle && keysInUse.count(keyOffset) == 0) {
             // Mark the unused key as deleted
             while (*keys != 0) { *keys++ = 1; }
             *keys++ = 1;
-            keyOffset = static_cast<int32_t>(keys - fKeys);
+        } else {
+            map[i].oldpos = keyOffset;
+            map[i].newpos = 0;
+            while (*keys != 0) { ++keys; }  /* skip the key */
+            ++keys;  /* skip the NUL */
+            i++;
         }
-        map[i].oldpos = keyOffset;
-        map[i].newpos = 0;
-        while (*keys != 0) { ++keys; }  /* skip the key */
-        ++keys;  /* skip the NUL */
     }
-    U_ASSERT(keys == fKeys + fKeysTop);
+    if (keys != fKeys + fKeysTop) {
```

Comment:
Optional: You could do this unconditionally.

Reply:
Acknowledged